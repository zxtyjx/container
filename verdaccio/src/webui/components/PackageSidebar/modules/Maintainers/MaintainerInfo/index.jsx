import React from 'react';
import PropTypes from 'prop-types';

import classes from './style.scss';

export default function MaintainerInfo({title, name, avatar}) {
  let avatarDescription = `${title} ${name}'s avatar`;
  return (
    <div className={classes.maintainer} title={name}>
      <img src={avatar} alt={avatarDescription} title={avatarDescription}/>
      <span className="maintainer-name">{name}</span>
    </div>
  );
}
MaintainerInfo.propTypes = {
  title: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  avatar: PropTypes.string.isRequired
};
