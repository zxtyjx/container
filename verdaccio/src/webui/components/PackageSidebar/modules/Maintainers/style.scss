@import '../../../../styles/variable';

.maintainersModule {
  .showAllContributors {
    cursor: pointer;
    width: 100%;
    background: none;
    border: none;
    font-size: 14px;
    text-align: center;
    padding: 10px 0;
  }
}
