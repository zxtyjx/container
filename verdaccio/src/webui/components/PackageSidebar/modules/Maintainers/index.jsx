import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash/get';
import filter from 'lodash/filter';
import size from 'lodash/size';
import has from 'lodash/has';
import uniqBy from 'lodash/uniqBy';

import Module from '../../Module';
import MaintainerInfo from './MaintainerInfo';
import ModuleContentPlaceholder from '../../ModuleContentPlaceholder';

import classes from './style.scss';

const CONTRIBUTORS_TO_SHOW = 5;

export default class Maintainers extends React.Component {
  static propTypes = {
    packageMeta: PropTypes.object.isRequired
  };

  state = {};

  constructor(props) {
    super(props);
    this.handleShowAllContributors = this.handleShowAllContributors.bind(this);
  }

  get author() {
    return get(this, 'props.packageMeta.latest.author');
  }

  get contributors() {
    let contributors = get(this, 'props.packageMeta.latest.contributors', {});
    return filter(contributors, (contributor) => {
      return (
        contributor.name !== get(this, 'author.name') &&
        contributor.email !== get(this, 'author.email')
      );
    });
  }

  get showAllContributors() {
    return this.state.showAllContributors || size(this.contributors) <= 5;
  }

  get uniqueContributors() {
    if (!this.contributors) {
      return [];
    }

    return uniqBy(this.contributors, (contributor) => contributor.name).slice(
      0,
      CONTRIBUTORS_TO_SHOW
    );
  }

  handleShowAllContributors() {
    this.setState({
      showAllContributors: true
    });
  }

  renderContributors() {
    if (!this.contributors) return null;

    return (this.showAllContributors
      ? this.contributors
      : this.uniqueContributors
    ).map((contributor, index) => {
      return (
        <MaintainerInfo
          key={index}
          title="Contributors"
          name={contributor.name}
          avatar={contributor.avatar}
        />
      );
    });
  }

  renderAuthorAndContributors(author) {
    return (
      <div>
        <ul className="maintainer-author">
          {author &&
            author.name && (
              <MaintainerInfo
                title="Author"
                name={author.name}
                avatar={author.avatar}
              />
            )}
          {this.renderContributors()}
        </ul>
        {!this.showAllContributors && (
          <button
            onClick={this.handleShowAllContributors}
            className={classes.showAllContributors}
            title="Current list only show the author and first 5 contributors unique by name"
          >
            Show all contributor
          </button>
        )}
      </div>
    );
  }

  render() {
    let author = this.author;
    const contributors = this.renderContributors();
    return (
      <Module title="Maintainers" className={classes.maintainersModule}>
        {contributors.length || has(author, 'name') ? (
          this.renderAuthorAndContributors(author)
        ) : (
          <ModuleContentPlaceholder text="Not Available!" />
        )}
      </Module>
    );
  }
}
