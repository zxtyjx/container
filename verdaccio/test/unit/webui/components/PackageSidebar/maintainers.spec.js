/**
 * Maintainers component
 */

import React from 'react';
import { mount } from 'enzyme';
import Maintainers from '../../../../../src/webui/components/PackageSidebar/modules/Maintainers/index';
import { packageMeta } from '../store/packageMeta';

describe('<PackageSidebar /> : <Maintainers />', () => {
  let wrapper;
  let instance;

  beforeEach(() => {
    wrapper = mount(<Maintainers packageMeta={packageMeta} />);
    instance = wrapper.instance();
  });

  it('should match with the props', () => {
    expect(wrapper.props().packageMeta).toEqual(packageMeta);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('author shoule be equal to User NPM', () => {
    expect(instance.author).toEqual({
      avatar:
        'https://www.gravatar.com/avatar/a5a236ba477ee98908600c40cda74f4a',
      email: 'test@author.local',
      name: 'User NPM'
    });
  });

  it('should get all the contributors with false for showAllContributors', () => {
    expect(instance.showAllContributors).toBeFalsy();
  });

  it('should get unique contributors', () => {
    const result = [
      {
        avatar:
          'https://www.gravatar.com/avatar/4ef03c2bf8d8689527903212d96fb45b',
        email: 'test1@test.local',
        name: '030'
      },
      {
        avatar:
          'https://www.gravatar.com/avatar/06975001f7f2be7052bcf978700c6112',
        email: 'tes4@test.local',
        name: 'Alex Vernacchia'
      },
      {
        avatar:
          'https://www.gravatar.com/avatar/d9acfc4ed4e49a436738ff26a722dce4',
        email: 'test5@test.local',
        name: 'Alexander Makarenko'
      },
      {
        avatar:
          'https://www.gravatar.com/avatar/2e095c7cfd278f72825d0fed6e12e3b1',
        email: 'test6@test.local',
        name: 'Alexandre-io'
      },
      {
        avatar:
          'https://www.gravatar.com/avatar/371edff6d79c39bb9e36bde39d41a4b0',
        email: 'test7@test.local',
        name: 'Aram Drevekenin'
      }
    ];
    expect(instance.uniqueContributors).toEqual(result);
  });

  it('should click on handleShowAllContributors', () => {
    wrapper.find('button').simulate('click');
    expect(wrapper.state('showAllContributors')).toBeTruthy();
  });
});
