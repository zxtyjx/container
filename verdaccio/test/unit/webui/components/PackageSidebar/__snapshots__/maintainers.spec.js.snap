// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PackageSidebar /> : <Maintainers /> should match with the props 1`] = `"<div class=\\"module maintainersModule\\"><h2 class=\\"moduleTitle\\">Maintainers</h2><div><div><ul class=\\"maintainer-author\\"><div class=\\"maintainer\\" title=\\"User NPM\\"><img src=\\"https://www.gravatar.com/avatar/a5a236ba477ee98908600c40cda74f4a\\" alt=\\"Author User NPM's avatar\\" title=\\"Author User NPM's avatar\\"><span class=\\"maintainer-name\\">User NPM</span></div><div class=\\"maintainer\\" title=\\"030\\"><img src=\\"https://www.gravatar.com/avatar/4ef03c2bf8d8689527903212d96fb45b\\" alt=\\"Contributors 030's avatar\\" title=\\"Contributors 030's avatar\\"><span class=\\"maintainer-name\\">030</span></div><div class=\\"maintainer\\" title=\\"Alex Vernacchia\\"><img src=\\"https://www.gravatar.com/avatar/06975001f7f2be7052bcf978700c6112\\" alt=\\"Contributors Alex Vernacchia's avatar\\" title=\\"Contributors Alex Vernacchia's avatar\\"><span class=\\"maintainer-name\\">Alex Vernacchia</span></div><div class=\\"maintainer\\" title=\\"Alexander Makarenko\\"><img src=\\"https://www.gravatar.com/avatar/d9acfc4ed4e49a436738ff26a722dce4\\" alt=\\"Contributors Alexander Makarenko's avatar\\" title=\\"Contributors Alexander Makarenko's avatar\\"><span class=\\"maintainer-name\\">Alexander Makarenko</span></div><div class=\\"maintainer\\" title=\\"Alexandre-io\\"><img src=\\"https://www.gravatar.com/avatar/2e095c7cfd278f72825d0fed6e12e3b1\\" alt=\\"Contributors Alexandre-io's avatar\\" title=\\"Contributors Alexandre-io's avatar\\"><span class=\\"maintainer-name\\">Alexandre-io</span></div><div class=\\"maintainer\\" title=\\"Aram Drevekenin\\"><img src=\\"https://www.gravatar.com/avatar/371edff6d79c39bb9e36bde39d41a4b0\\" alt=\\"Contributors Aram Drevekenin's avatar\\" title=\\"Contributors Aram Drevekenin's avatar\\"><span class=\\"maintainer-name\\">Aram Drevekenin</span></div></ul><button class=\\"showAllContributors\\" title=\\"Current list only show the author and first 5 contributors unique by name\\">Show all contributor</button></div></div></div>"`;
